truffleruby/src/main/java/org/truffleruby/core/time/TimeNodes.java

```diff
@@ -332,16 +338,40 @@ public DynamicObject timeDecompose(DynamicObject time) {
     }
 
     @Primitive(name = "time_strftime")
+    @ImportStatic(StringCachingGuards.class)
     public static abstract class TimeStrftimePrimitiveNode extends PrimitiveArrayArgumentsNode {
 
-        @TruffleBoundary
+        @Child private GetTimeZoneNode getTimeZoneNode = GetTimeZoneNodeGen.create();
+        
+        @Specialization(guards = { "isRubyString(format)",
+                                   "ropesEqual(format, cachedFormat)" } )
+        public DynamicObject timeStrftime(VirtualFrame frame, DynamicObject time, DynamicObject format,
+                                          @Cached("privatizeRope(format)") Rope cachedFormat,
+                                          @Cached("createFormatter()") RubyDateFormatter rdf,
+                @Cached("compilePattern(rdf, cachedFormat)") List<Token> pattern) {
+            return createString(format(rdf, pattern, Layouts.TIME.getDateTime(time), getTimeZoneNode.executeGetTimeZone(frame)));
+        }
+
         @Specialization(guards = "isRubyString(format)")
-        public DynamicObject timeStrftime(DynamicObject time, DynamicObject format) {
-            final RubyDateFormatter rdf = new RubyDateFormatter(getContext(), this);
-            final List<Token> pattern = rdf.compilePattern(StringOperations.rope(format), false);
-            return createString(rdf.formatToRopeBuilder(pattern, Layouts.TIME.getDateTime(time)));
+        public DynamicObject timeStrftime(VirtualFrame frame, DynamicObject time, DynamicObject format) {
+            final  RubyDateFormatter rdf = new RubyDateFormatter(getContext(), this);
+            final List<Token> pattern = compilePattern(rdf, StringOperations.rope(format));
+            return createString(format(rdf, pattern, Layouts.TIME.getDateTime(time), getTimeZoneNode.executeGetTimeZone(frame)));
+        }
+
+        protected RubyDateFormatter createFormatter() {
+            return new RubyDateFormatter(getContext(), this);
```

Comment:
The RubyDateFormatter should stateless and singleton, and context/currentNode be passed to the methods.
OTOH the lexer doesn't seem stateless.
But that's for another time, I'll do after this PR.